emcache/cluster.py

```diff
@@ -221,6 +273,93 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
 
         raise ValueError(f"{memcached_host_address} can not be found")
 
+    async def _autodiscovery_monitor(self) -> None:
+        logger.debug("Autodiscovery task started")
+        while True:
+            try:
+                if await self.autodiscover():
+                    await asyncio.sleep(self._autodiscovery_poll_interval)
+                else:
+                    await asyncio.sleep(AUTODISCOVERY_POLL_INTERVAL_FAILURE)
+            except asyncio.CancelledError:
+                # if a cancellation is received we stop monitoring for cluster changes
+                break
+
+            except Exception:
+                logger.exception("Autodiscover raised an exception, continuing processing")
+
+        logger.debug("Autodiscovery task stopped")
+
+    async def _get_autodiscovered_nodes(self) -> tuple[bool, int, list[tuple[str, str, int]]]:
+        if self._closed:
+            raise RuntimeError("Emcache client is closed")
+
+        node = random.choice(self._discovery_nodes if self._discovery_nodes else self.nodes)
+        async with OpTimeout(self._autodiscovery_timeout, self._loop):
+            async with node.connection() as connection:
+                return await connection.autodiscovery()
+
+    async def autodiscover(self) -> bool:
+        try:
+            autodiscovery, version, nodes = await self._get_autodiscovered_nodes()
+        except asyncio.TimeoutError as e:
+            logger.error("Got timeout when checking cluster configuration: %r", e)
+            return False
+
+        if not autodiscovery:
+            logger.warning("autodiscovery request failed, does the server support 'config get cluster' command?")
+            return False
+
+        if version == self._autodiscover_config_version:
+            return False
+
+        if not nodes:
+            logger.error("Node list received from autodiscovery is empty!")
+            return False
+
+        # Updating node list with preservation of their status
+        new_nodes_set = {MemcachedHostAddress(host, port) for host, ip, port in nodes}
+        old_nodes_set = {node.memcached_host_address for node in self.nodes}
+        protected = {node.memcached_host_address for node in self._discovery_nodes}
```

Comment:
Is this something really necessary? I would expect - please correct me if Im wrong - that AWS provides you a generic hostname for making the discovery which could resolve to any host of the cluster, while the response returned back from the command for making the discovery would return exclusively hostnames that can resolve to the specific node hosts.

If this was true, I would expect that we do not need to make this protection.

A related conversation, is how the usage of a connection pool can impact the discovery? Remember that in the end when we instantiate a Node we are creating a connection pool where here is 1 or multiple connections already opened. Worst case scenario could be tens of them. If one of the nodes of the cluster is removed and we had the back luck of making the resolution to that node, the connection pool could hold invalid TCP connections which would imply that the autodiscovery command would fail. True that eventually all TCP connections would get recycled and eventually new DNS resolutions and new TCP connections will be created.

Wondering if we should be an step ahead on this scenario, we should double check this, and avoid this situation by avoiding the connection pool usage or just use another strategy like making the discovery using all healthy nodes until one succeed.

Thoughts?

Reply:
Actually this was a problem when I placed that generic hostname in `self._healthy_nodes` as this is no longer true, this code might indeed not be neeeded anymore.

Reply:
removed that code